src/client/components/Buttons/Button.tsx

```diff
@@ -0,0 +1,334 @@
+		justify-content: flex-start;
+	}
+`;
+
+const getIcon = (
+	IconProp: ButtonBase['icon'],
+	iconAlt: ButtonBase['iconAlt']
+): JSX.Element | null => {
+	let Icon: JSX.Element | null = null;
+	if (IconProp) {
+		if (typeof IconProp === 'string') {
+			const ButtonIcon: FC = () => <img src={IconProp} alt={iconAlt} />;
+			Icon = <ButtonIcon />;
+		} else {
+			Icon = <IconProp />;
+		}
+	}
+	return Icon;
+};
+
+const getClickHandler = (
+	props: ButtonProps,
+	setUnresolved: React.Dispatch<React.SetStateAction<boolean>>
+): ButtonProps['onClick'] => {
+	if (props.async) {
+		// TypeScript can infer that onClick is set because of the `props.async`
+		// type guard, but cannot guarantee that it'll be set inside the closure,
+		// but capturing the reference here satisfies that guarantee.
+		const delegatedClickHandler = props.onClick;
+		return () => {
+			const result = delegatedClickHandler();
+			setUnresolved(true);
+			setTimeout(async () => {
+				// Wait 700 ms or until the action completes, whichever comes later.
+				// This prevents the quick flash of the loader, while also providing
+				// feedback to the user.
+				await result;
+				setUnresolved(false);
+			}, 700);
+		};
+	}
+	return props.onClick;
+};
+
+const getClassNames = (props: ButtonProps): string => {
+	const classNames: string[] = [];
+	if (props.outline) classNames.push('outline');
+	else classNames.push('filled');
+	if (props.small) classNames.push('small');
+	if (props.large) classNames.push('large');
+	if (props.loading) classNames.push('loading');
+	if (props.long) classNames.push('long');
+	if (props.alignStart) classNames.push('align-start');
+	if (props.secondary) classNames.push('secondary');
+	if (props.disabled) classNames.push('disabled');
+	if (props.warning) classNames.push('warning');
+	return classNames.join(' ');
+};
+
+export const Button: FC<ButtonProps> = props => {
+	const [unresolved, setUnresolved] = useState(false);
+	const { icon, iconAlt } = props;
+	const iconElement = useMemo(() => getIcon(icon, iconAlt), [icon, iconAlt]);
+	const clickHandler = useMemo(() => getClickHandler(props, setUnresolved), [props]);
+	const classNames = useMemo(() => getClassNames(props), [props]);
+	const { children, type = 'button', linkTo, secondary, async, loading } = props;
+	const buttonElement = (
+		<StyledButton
+			onClick={clickHandler}
+			className={`${classNames}${unresolved ? ' loading' : ''}`}
+			role="button"
+			type={type}
+			secondary={secondary}>
+			<div className="text">
+				{iconElement}
+				{children}
+			</div>
+			{async || loading ? <Spinner /> : null}
+		</StyledButton>
+	);
+
+	// Rather than handle in onClick, native DOM elements have better a11y support.
+	if (linkTo) {
+		if (props.externalLink || isAbsoluteUrl(linkTo)) {
+			return (
+				<a rel="noopener" href={linkTo}>
+					{buttonElement}
+				</a>
+			);
+		}
+		return <Link to={linkTo}>{buttonElement}</Link>;
+	}
+	return buttonElement;
+};
+
+export default Button;
```

Comment:
We don't have to settle this here but... my personal opinion is that default exports are the devil and shouldn't be used.

Reply:
Agreed.